Add notification modal for missing ratings

PatricKangasmaki opened the pull request:
Notifications modal for users that haven't given ratings when they should (it parses out logged in user, I figured that you don't want to send email to yourself). 
Currently it sends the api request to missing endpoint so it should give 404. 
Payload contains: List of users that were checked and the correct task.

IidaKainu commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,225 @@
+      return setErrorMessage(t('Please select at least one user'));
+    }
+
+    dispatch(roadmapsActions.notifyUsers({ users: usersToNotify, task })).then(
+      (res) => {
+        setIsLoading(false);
+        if (roadmapsActions.notifyUsers.rejected.match(res)) {
+          if (res.payload) {
+            setErrorMessage(res.payload.message);
+          }
+        } else {
+          closeModal();
+        }
+      },
+    );
```
The readability of this could be improved with async/await
```suggestion
    const res = await dispatch(
      roadmapsActions.notifyUsers({ users: usersToNotify, task }),
    );
    setIsLoading(false);
    if (roadmapsActions.notifyUsers.rejected.match(res)) {
      if (res.payload) setErrorMessage(res.payload.message);
      return;
    }
    closeModal();
```

IidaKainu commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,225 @@
+      checkedUsers?.some((checkedUser) => checkedUser.id === user.id),
+    );
+
+    if (!usersToNotify || usersToNotify?.length === 0) {
+      setIsLoading(false);
+      return setErrorMessage(t('Please select at least one user'));
```
Instead of showing an error when no users are selected, the submit button could be disabled.

IidaKainu commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,225 @@
+          <Checkbox
+            label="All"
+            checked={allChecked}
+            onChange={(checked: boolean) => checkAll(checked)}
```
The types on the checkbox listeners can be left out due to type inference
```suggestion
            onChange={(checked) => checkAll(checked)}
```

IidaKainu commented on frontend/src/api/api.tsx:
```diff
@@ -305,6 +305,11 @@ const patchIntegrationConfiguration = async (
 };
 
+const sendNotification = async (users: RoadmapUser[], task: Task) => {
+  const response = await axios.post(`/notify`, { users, task });
```
I'm wondering if the route should follow the pattern of roadmaps/tasks/notify?
```suggestion
  const response = await axios.post(
    `roadmaps/${task.roadmapId}/tasks/${task.id}/notify`,
    { users, task },
  );
```

IidaKainu commented on frontend/src/redux/roadmaps/actions.ts:
```diff
@@ -340,3 +340,16 @@ export const patchIntegrationConfiguration = createAsyncThunk<
 );
+
+export const notifyUsers = createAsyncThunk<
+  boolean,
```
I think return value should be void. The response status code will tell whether the operation was successful.
```suggestion
  void,
```

IidaKainu reviewed (changes requested):
The notified users look correct
The "all" -checkbox behaves really intuitively, great work 👍

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,225 @@
+
+  const checkUser = (checked: boolean, idx: number) => {
+    setAllChecked(false);
+    const copy = [...(missingUsers ?? [])];
+    if (!copy) return;
```
`copy` will never be undefined here, so the condition is always false (same above in `checkAll`)

I think I saw the same mistake in some other place as well, outside this pr

```suggestion
    if (!missingUsers) return;
    const copy = [...missingUsers];
```

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,225 @@
+        .map((dev) => ({ ...dev, checked: false }));
+
+      setMissingUsers([
+        ...(missingRepresentatives ?? []),
```
won't be undefined
```suggestion
        ...missingRepresentatives,
```

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,225 @@
+          if (representatives) {
+            representatives.forEach((rep) => {
+              if (!ratingsForTask.includes(rep.id))
+                if (!missingRepresentatives.some((e) => e.id === rep.id))
+                  missingRepresentatives.push({ ...rep, checked: false });
```
`Map<id, user>` could be used, instead of an array, for the deduplication (not that this would be a huge problem any time soon)

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,225 @@
+        createdByUser: rating.createdByUser,
+        forCustomer: rating.forCustomer,
+      }));
+      const missingCustomers = customers?.filter((customer) => {
+        const representativeIds = customer?.representatives?.map(
+          (rep) => rep.id,
```
Could these use the same functions used for the task listing page? (just filtering out the current user)

juhoeerola commented:
> The "all" -checkbox behaves really intuitively, great work 👍

^ this

One possibly nice to have behaviour, would be remembering the individual selection separate to "select all", so that toggling "check all" twice would be idempotent.

Then you would lose the ability to "uncheck all" with that, though... (that would require manual unchecking of every manually checked user)

PatricKangasmaki commented:
Added all the suggestions!


> Additionally, it might be nice to have ability to include an optional message to the notification, which would be added to the email.

I added the possibility to send messages in theory but didn't add input fields for it yet, I think we should consult our designer on that part.

> One possibly nice to have behaviour, would be remembering the individual selection separate to "all", so that toggling "all" twice would be idempotent.

The checkboxes now save the previous state so toggling back from "all" will have the previously selected ones, big thanks to Iida here!

juhoeerola commented on frontend/src/utils/TaskUtils.ts:
```diff
@@ -383,3 +383,33 @@ export const unratedProductOwnerTasks: (
+    const representativeIds = customer?.representatives?.map((rep) => rep.id);
+    return !representativeIds?.every((rep) => {
+      return ratings.some((rating) => {
+        if (rating.createdByUser === rep && rating.forCustomer === customer.id)
+          return true;
+        return false;
```
Could just return the value of the condition
```suggestion
        return rating.createdByUser === rep && rating.forCustomer === customer.id;
```

juhoeerola commented on frontend/src/utils/TaskUtils.ts:
```diff
@@ -383,3 +383,33 @@ export const unratedProductOwnerTasks: (
+    forCustomer: rating.forCustomer,
+  }));
+  return customers?.filter((customer) => {
+    const representativeIds = customer?.representatives?.map((rep) => rep.id);
+    return !representativeIds?.every((rep) => {
+      return ratings.some((rating) => {
+        if (rating.createdByUser === rep && rating.forCustomer === customer.id)
```
Here also mapping out the `id`s could be skipped
```suggestion
    return !customer.representatives?.every((rep) => {
      return ratings.some((rating) => {
        if (rating.createdByUser === rep.id && rating.forCustomer === customer.id)
```

And `customer` won't be undefined

juhoeerola commented on frontend/src/utils/TaskUtils.ts:
```diff
@@ -383,3 +383,33 @@ export const unratedProductOwnerTasks: (
+) => {
+  const ratingIds = taskRatings.map((rating) => rating.createdByUser);
+  return allUsers
+    ?.filter((user) => user.type === RoleType.Developer)
+    .filter((developer) => !ratingIds.includes(developer.id));
```
The filters could be combined, and `allUsers` should not be undefined
```suggestion
    .filter((user) => user.type === RoleType.Developer && !ratingIds.includes(user.id));
```

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,223 @@
+              <button
+                className="button-large"
+                type="submit"
+                disabled={!missingUsers?.filter((user) => user.checked).length}
```
```suggestion
                disabled={!missingUsers?.some((user) => user.checked)}
```

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,223 @@
+  };
+
+  return (
+    <>
```
Unnecessary fragment, as the single `Form` element is returned
```suggestion
```

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,223 @@
+    if (checked) {
+      const copy = [...missingUsers];
+      setPreviousUsers(copy);
+      setMissingUsers(copy.map((_user) => ({ ..._user, checked })));
```
The underscore should be dropped, as it usually indicates unused parameter
```suggestion
      setMissingUsers(copy.map((user) => ({ ...user, checked })));
```

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,223 @@
+      ).map((dev) => ({ ...dev, checked: false }));
+
+      setMissingUsers([
+        ...(Array.from(missingRepresentatives.values()) ?? []),
+        ...(missingDevelopers ?? []),
```
Neither will be undefined
```suggestion
        ...Array.from(missingRepresentatives.values()),
        ...missingDevelopers,
```

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,223 @@
+      if (missingCustomers) {
+        missingCustomers.forEach((customer) => {
+          // Filter out logged in user, so they won't send email to themselves
+          const representatives = customer?.representatives?.filter(
```
also, customer won't be undefined
```suggestion
          const representatives = customer.representatives?.filter(
```

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,223 @@
+      const missingRepresentatives = new Map<number, CheckableUser>();
+
+      // Find representatives who haven't given ratings
+      if (missingCustomers) {
```
`missingCustomers` won't be false (if the types are correct for `findMissingCustomers`)
```suggestion
```

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,223 @@
+    if (allChecked) setAllChecked(false);
+    const copy = [...missingUsers];
+    copy[idx].checked = checked;
+    setMissingUsers(copy);
```
A 🍒 on top could be setting `allChecked` if all are checked manually
```suggestion
    setMissingUsers(copy);
    setAllChecked(checked && copy.every((user) => user.checked));
```

juhoeerola commented on frontend/src/components/modals/NotifyUsersModal.tsx:
```diff
@@ -0,0 +1,223 @@
+            representatives.forEach((rep) => {
+              if (!ratingsForTask.includes(rep.id))
+                if (
+                  !Array.from(missingRepresentatives.values()).some(
+                    (e) => e.id === rep.id,
+                  )
```
As far as I could see, this is checking:
```suggestion
                  !missingRepresentatives.has(rep.id)
```

(the nested if conditions could also be combined to one)

juhoeerola reviewed (changes requested):
Seems to work great 👍 

Comments mostly on small simplifications
There were also quite a few defensive uses of `?.`, where the type involved didn't include `null` or `undefined`

PatricKangasmaki commented:
All the suggestions you gave are now fixed and I've added the new input field for the optional message 👍

IidaKainu reviewed (approved):
I went over the suggestions and the changes look good, as does the input field. 
Just a small thing I found: the label 'Optional message' is not in translations. Otherwise, I'm giving my approval 👍
